go/vt/discovery/healthcheck.go

```diff
@@ -548,41 +616,23 @@ func (hcc *healthCheckConn) setServingState(serving bool, reason string) {
 
 // stream streams healthcheck responses to callback.
 func (hcc *healthCheckConn) stream(ctx context.Context, hc *HealthCheckImpl, callback func(*querypb.StreamHealthResponse) error) {
-	hcc.mu.Lock()
-	conn := hcc.conn
-	hcc.mu.Unlock()
-
-	if conn == nil {
-		var err error
-		conn, err = tabletconn.GetDialer()(hcc.tabletStats.Tablet, grpcclient.FailFast(true))
+	if hcc.conn == nil {
+		conn, err := tabletconn.GetDialer()(hcc.tabletStats.Tablet, grpcclient.FailFast(true))
 		if err != nil {
-			hcc.mu.Lock()
 			hcc.tabletStats.LastError = err
-			hcc.mu.Unlock()
 			return
 		}
-
-		hcc.mu.Lock()
 		hcc.conn = conn
 		hcc.tabletStats.LastError = nil
-		hcc.mu.Unlock()
 	}
 
-	if err := conn.StreamHealth(ctx, callback); err != nil {
-		hcc.mu.Lock()
+	if err := hcc.conn.StreamHealth(ctx, callback); err != nil {
+		hcc.conn.Close(ctx)
```

Comment:
This is a somewhat dangerous and subtle reintroduction of the same flavor of bug that blacked out all of Slack's vtgates for some time....

We need to move this down below to make sure that we notify hc that the tablet is not serving _before_ trying to close the connection. Since Close can be a slow over-the-network operation, we don't want the tablet record in serving state when it's not actually viable.

Granted, we no longer have that same lock inversion problem which caused the deadlock with this newer lock-free design, but I still think the reordering is important.

Reply:
Makes sense. Done.